editor/utils/heartbeat.js

```diff
@@ -0,0 +1,173 @@
+import store from '../store';
+import {
+	getCurrentPost,
+	getEditedPostTitle,
+	getEditedPostExcerpt,
+	getEditedPostContent,
+} from '../store/selectors';
+import {
+	toggleAutosave,
+	removeNotice,
+	toggleNetworkIsConnected,
+	showDisconnectionNotice,
+} from '../store/actions';
+
+import { compact } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { doAction } from '@wordpress/hooks';
+
+const DISCONNECTION_NOTICE_ID = 'DISCONNECTION_NOTICE_ID';
+
+export function setupHeartbeat() {
+	const $document = jQuery( document );
+	/**
+	 * Configure heartbeat to refresh the wp-api nonce, keeping the editor authorization intact.
+	 *
+	 * @todo update the _wpnonce used by autosaves.
+	 */
+	$document.on( 'heartbeat-tick', ( event, response ) => {
+		if ( response[ 'rest-nonce' ] ) {
+			window.wpApiSettings.nonce = response[ 'rest-nonce' ];
+		}
+	} );
+
+	/**
+	 * Concatenate the title, content and excerpt.
+	 *
+	 * This is used to track changes when auto-saving.
+	 *
+	 * @since 1.9.0
+	 *
+	 * @param {Object} state The current state.
+	 *
+	 * @returns {string} A concatenated string with title, content and excerpt.
+	 */
+	const getCompareString = function( state ) {
+		return compact( [
+			getEditedPostTitle( state ),
+			getEditedPostContent( state ),
+			getEditedPostExcerpt( state ),
+		] ).join( '::' );
+	};
+
+	/**
+	 * Configure Heartbeat autosaves.
+	 */
+	const { dispatch, getState } = store;
+
+	let compareString;
+	let lastCompareString;
+
+	/**
+	 * Autosave 'save' function that pulls content from Gutenberg state. Based on `wp.autosave.save`.
+	 *
+	 * @since 1.9.0
+	 *
+	 * @returns {Object|bool} postData The autosaved post data to send, or false if no autosave is needed.
+	 */
+	const save = function() {
+		// Bail early if autosaving is suspended or saving is blocked.
+		if ( wp.autosave.isSuspended || wp.autosave._blockSave ) {
+			return false;
+		}
+
+		// Check if its time for another autosave.
+		if ( ( new Date() ).getTime() < wp.autosave.nextRun ) {
+			return false;
+		}
+
+		// Get the current editor state and compute the compare string (title::excerpt::content).
+		const state = getState();
+		compareString = getCompareString( state );
+
+		// Initialize lastCompareString if this is the first run and it is undefined.
+		if ( typeof lastCompareString === 'undefined' ) {
+			lastCompareString = compareString;
+		}
+
+		// Bail if no changes to the compare string (title::excerpt::content).
+		if ( compareString === lastCompareString ) {
+			return false;
+		}
+
+		// Store the latest compare string.
+		lastCompareString = compareString;
+
+		// Block autosaving for 10 seconds.
+		wp.autosave.server.tempBlockSave();
+
+		const postData = getCurrentPost( state );
```

Comment:
I believe this returns the last saved post and if I understand correctly we want to save the last edits here which are not included in that object. What am I missing here?

Reply:
oh, thats a mistake then - this should be getting the current editor content; what should I be calling to get that?

Reply:
something like that 

```
		const post = getCurrentPost( state );
		const edits = getPostEdits( state );
		const toSend = {
			...edits,
			content: getEditedPostContent( state ),
			id: post.id,
		};
```

see the `REQUEST_POST_UPDATE` effect

Reply:
I simplified the data setup here in https://github.com/WordPress/gutenberg/pull/4218/commits/158bbd4b4a686f67003d43e7bdb794d608a730e9, we don't need the full post object for the autosave.